### PR TITLE
Lazy `Term` visitor

### PR DESCRIPTION
**Pull request by wsx-ucb**
Cherry-picked a commit from `future` branch

### CODE REVIEW

File: base/src/main/java/org/aya/core/visitor/TermPostMap.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.core.visitor;
+
+import org.aya.core.term.Term;
+
+import java.util.function.Function;
+
+public record TermPostMap(TermView view, Function<Term, Term> f) implements TermView {
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/TermPostMap.java:9]: Isn't this just `TermMap` but weaker?

File: base/src/main/java/org/aya/core/visitor/TermPreMap.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.core.visitor;
+
+import org.aya.core.term.Term;
+
+import java.util.function.Function;
+
+public record TermPreMap(TermView view, Function<Term, Term> f) implements TermView {
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/TermPreMap.java:9]: Ditto

**Review by ice1000 — COMMENTED** [review]
Remember to add not-null annotations and check CI.

I can confirm that the idea behind the code in `commit` overloads are mostly what I asked. I'll have to check more carefully to see if they're really correct though.

Great job!!

File: base/src/main/java/org/aya/core/visitor/TermPostMap.java (hunk 1 above)
Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/visitor/TermPostMap.java:9]: Yes, this is just `TermPostMap(view, id, f)`.
Not sure if we need this just for convenience.

**Comment by wsx-ucb** [comment]
> I can confirm that the idea behind the code in commit overloads are mostly what I asked. I'll have to check more carefully to see if they're really correct though.

The `note/unfiled.md` has some information on the semantics.
The `traverse` function just follows the formalism there.
Also, the subtle conditions of composing pre/post functions are provided in the last paragraph.

**Comment by ice1000** [comment]
> > I can confirm that the idea behind the code in commit overloads are mostly what I asked. I'll have to check more carefully to see if they're really correct though.
> 
> The `note/unfiled.md` has some information on the semantics. The `traverse` function just follows the formalism there. Also, the subtle conditions of composing pre/post functions are provided in the last paragraph.

Yeah!! The document is greeeaaaat

**Review by imkiva — APPROVED** [review]
Wow, the document is so good!!

File: base/src/main/java/org/aya/core/visitor/TermPostMap.java (hunk 1 above)
Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/visitor/TermPostMap.java]: I removed the `TermPreMap` and `TermPostMap` records, but still keep the methods.

Comment (ice1000) [reply base/src/main/java/org/aya/core/visitor/TermPostMap.java]: Great

File: base/src/main/java/org/aya/core/term/Term.java [hunk 3]
```diff
@@ -216,4 +216,8 @@ public static boolean checkSubst(@NotNull SeqLike<@NotNull Param> params, @NotNu
     }
   }
 
+  default TermView view() {
+    var term = this;
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/term/Term.java:220]: Why

Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/term/Term.java:220]: Good catch, I guess we can inline that.

**Comment by ice1000** [comment]
bros ping

**Comment by ice1000** [comment]
bors ping

**Comment by ice1000** [comment]
bros r+

**Pull request merged**